### PR TITLE
Add ByteArray Inlet/Outlet, fixed some deprecated messages

### PR DESCRIPTION
**Pull request by blublinsky**

### CODE REVIEW

**Comment by blublinsky** [comment]
Now with simple unit test

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayInlet.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  def schemaAsString                                                                                      = ByteArrayUtil.schemaName
+  def withUniqueGroupId: ByteArrayInlet                                                                   = copy(hasUniqueGroupId = true)
+  override def withErrorHandler(handler: (Array[Byte], Throwable) => Option[Array[Byte]]): ByteArrayInlet = copy(errorHandler = handler)
+}
+
+object ByteArrayInlet {
+  // Java API
+  def create(name: String): ByteArrayInlet = ByteArrayInlet(name)
+
+  def create(name: String, hasUniqueGroupId: Boolean): ByteArrayInlet = ByteArrayInlet(name, hasUniqueGroupId)
```
Comment (RayRoestenburg) [inline core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayInlet.scala:37]:
```suggestion
  // Java API
  def create(name: String, hasUniqueGroupId: Boolean): ByteArrayInlet = ByteArrayInlet(name, hasUniqueGroupId)
```

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import java.security.MessageDigest
+import java.util.Base64
+
+import cloudflow.streamlets.SchemaDefinition
+
+object ByteArrayUtil {
+  val Format     = "bytearray"
+  val schemaName = "ByteArraySchema"
+
+  def createSchemaDefinition() = SchemaDefinition(
```
Comment (RayRoestenburg) [inline core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala:28]: It might be interesting to add a schema-name to ByteArrayInlet / Outlet, to get like at least a 'nominal type' for byte arrays.
Doing it like here means any byte array inlet and byte array outlet can read  / write to a topic even if they do not match at all, which seems like a programming error waiting to happen.

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/proto/javadsl/ProtoInlet.scala [hunk 3]
```diff
@@ -33,7 +30,7 @@ final case class ProtoInlet[T <: GeneratedMessageV3](
   val descriptor = clazz.getMethod("getDescriptor").invoke(null).asInstanceOf[Descriptor]
 
   val codec            = new ProtoCodec[T](clazz)
-  def schemaAsString   = TextFormat.printToUnicodeString(descriptor.toProto)
+  def schemaAsString   = TextFormat.printer.escapingNonAscii(false).printToString(descriptor.toProto)
```
Comment (RayRoestenburg) [inline core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/proto/javadsl/ProtoInlet.scala:33]: Why is this needed? Dod you run into an error?

Comment (RayRoestenburg) [reply core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/proto/javadsl/ProtoInlet.scala:33]: Ah, I see:
```
Deprecated. Use printer().escapingNonAscii(false).printToString(MessageOrBuilder)
static java.lang.String
```
Good catch.

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala (hunk 2 above)
Comment (RayRoestenburg) [reply core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala:28]: Have you tested this for instance with a byte array inlet and an avro outlet connected  to the same topic? I think you will get a verification error in the blueprint. See `checkCompatibility` in `cloudflow.blueprint.Topic`

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc [hunk 4]
```diff
@@ -111,6 +111,21 @@ Here `CustomHandler` is any function adhering to the following interface
 errorHandler: (Array[Byte], Throwable) => Option[T]
 ----
 
+== ByteArray Inlets and outlets
+
+As mentioned earlier, Cloudflow uses a schema-first approach for building streaming data pipelines.
+This works great for intra streamlet communications, but often falls short for integration
+with existing applications. Cloudflow currently supports Avro and Protobuf schemas, which might be an
+issue in cases, when existing applications do not support either one of them. ByteArray inlets and outlets
+are introduced specifically to support such integrations.
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:120]: I think we should add some docs on how to write your own codecs. ByteArray is a fallback / catch all, but in many cases I would assume users don't actually want to read/write byte arrays, they want to use their own codecs, for "external use". maybe have some `ExternalInlet` / `ExternalOutlet` that can only be attached on one side of a topic (producer or consumer), and exclude these in verification of the blueprint. LMKWYT.

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc [hunk 5]
```diff
@@ -111,6 +111,21 @@ Here `CustomHandler` is any function adhering to the following interface
...
+
+NOTE: ByteArray inlets and outlets are not intended for intra streamlets communications. They should only
+be used for Kafka-based integration with existing applications supporting proprietary message formats,
+for example JSON
+
+ByteArray inlets and outlets use byte arrays for sending and recieving messages from Kafka and as a
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:126]:
```suggestion
ByteArray inlets and outlets use byte arrays for sending and receiving messages from Kafka and as a
```

**Review by RayRoestenburg — COMMENTED** [review]
Let me know wat you think about type-/schema-name for byte array outlets / inlets and the external in/out-let idea.

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc (hunk 4 above)
Comment (blublinsky) [reply docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:120]: As I was writing in the doc, I do not see this used for streamlet to streamlet communications. Its a way to integrate with other applications, that do not support Avro or Proto. I am not sure we should encourage writing custom codecs, especially for streamlet to streamlet. Its just a connivence method for custom integrations

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala (hunk 2 above)
Comment (blublinsky) [reply core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala:28]: Of course it will fail. The idea is to use bytearray as inlet/outlet on the topics, where non cloudflow app writing/reading

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc (hunk 4 above)
Comment (RayRoestenburg) [reply docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:120]: no I'm saying specifically for external communications.

Comment (RayRoestenburg) [reply docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:120]: `ExternalInlet`, `ExternalOutlet`, make that formal.

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala (hunk 2 above)
Comment (RayRoestenburg) [reply core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala:28]: Ok so this is then only useful for external inlets and outlets. It's probably handy, also for verification, to know that these are in fact external inlets and outlets.

**Comment by blublinsky** [comment]
OK, I renamed to external inlet/outlet. That's probably as much as I can do

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala (hunk 2 above)
Comment (blublinsky) [reply core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/bytearray/ByteArrayUtil.scala:28]: renamed inlet outlet and reworked documentation accordingly

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc [hunk 6]
```diff
@@ -111,6 +111,22 @@ Here `CustomHandler` is any function adhering to the following interface
 errorHandler: (Array[Byte], Throwable) => Option[T]
 ----
 
+== External Inlets and outlets
+
+As mentioned earlier, Cloudflow uses a schema-first approach for building streaming data pipelines.
+This works great for intra streamlet communications, but often falls short for integration
+with existing applications which might use their proprietary encoding for messages.
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:118]:
```suggestion
with existing applications which might use proprietary encoding for messages.
```

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc [hunk 7]
```diff
@@ -111,6 +111,22 @@ Here `CustomHandler` is any function adhering to the following interface
...
+To accomodate these use cases Cloudflow introduces `ExternalInlet` and `ExternalOutlet`,
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:119]:
```suggestion
To accommodate these use cases Cloudflow introduces `ExternalInlet` and `ExternalOutlet`,
```

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc [hunk 8]
```diff
@@ -111,6 +111,22 @@ Here `CustomHandler` is any function adhering to the following interface
...
 ----
 
+== External Inlets and outlets
+
+As mentioned earlier, Cloudflow uses a schema-first approach for building streaming data pipelines.
+This works great for intra streamlet communications, but often falls short for integration
+with existing applications which might use proprietary encoding for messages.
+To accomodate these use cases Cloudflow introduces `ExternalInlet` and `ExternalOutlet`,
+which implement Kafka communications in `ByteArray` format, allowing user to explicitely
+marshal/unmarshall data as port of the application code.
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:121]:
```suggestion
marshal/unmarshall data as part of the application code.
```

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc [hunk 9]
```diff
@@ -111,6 +111,22 @@ Here `CustomHandler` is any function adhering to the following interface
...
+To accommodate these use cases Cloudflow introduces `ExternalInlet` and `ExternalOutlet`,
+which implement Kafka communications in `ByteArray` format, allowing user to explicitely
+marshal/unmarshall data as port of the application code.
+
+When using `ExternalInlet` each data record is read by streamlet as an array of bytes.
+When using `ExternalOutlet` each data record should be converted to an array of bytes before
+writing it into an outlet.
+
+NOTE: `ExternalInlet` and `ExternalOutlet` are not intended for intra streamlets communications. They should only
+be used for Kafka-based integration with existing applications supporting proprietary message formats
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:128]:
```suggestion
be used for Kafka-based integration with existing applications supporting proprietary message formats.
```

**Review by RayRoestenburg — APPROVED** [review]

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc (hunk 4 above)
Comment (RayRoestenburg) [reply docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc]: But we can do that , if users ask for it, in a separate PR. Right now this is good enough.

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc [hunk 10]
```diff
@@ -111,6 +111,22 @@ Here `CustomHandler` is any function adhering to the following interface
 errorHandler: (Array[Byte], Throwable) => Option[T]
 ----
 
+== External Inlets and outlets
+
+As mentioned earlier, Cloudflow uses a schema-first approach for building streaming data pipelines.
+This works great for intra streamlet communications, but often falls short for integration
+with existing applications which might use proprietary encoding for messages.
+To accommodate these use cases Cloudflow introduces `ExternalInlet` and `ExternalOutlet`,
+which implement Kafka communications in `ByteArray` format, allowing user to explicitely
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:120]:
```suggestion
which implement Kafka communications in `ByteArray` format, allowing user to explicitly
```

Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:120]:
```suggestion
which implement Kafka communications in `ByteArray` format, allowing user to explicitly
```

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc [hunk 11]
```diff
@@ -111,6 +111,22 @@ Here `CustomHandler` is any function adhering to the following interface
...
+marshal/unmarshall data as part of the application code.
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:121]:
```suggestion
marshal/unmarshal data as part of the application code.
```

File: docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc [hunk 12]
```diff
@@ -111,6 +111,22 @@ Here `CustomHandler` is any function adhering to the following interface
...
+with existing applications which might use proprietary encoding for messages.
+To accommodate these use cases Cloudflow introduces `ExternalInlet` and `ExternalOutlet`,
+which implement Kafka communications in `ByteArray` format, allowing user to explicitly
+marshal/unmarshall data as part of the application code.
+
+When using `ExternalInlet` each data record is read by streamlet as an array of bytes.
+When using `ExternalOutlet` each data record should be converted to an array of bytes before
+writing it into an outlet.
+
+NOTE: `ExternalInlet` and `ExternalOutlet` are not intended for intra streamlets communications. They should only
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/schema-first-approach.adoc:127]:
```suggestion
NOTE: `ExternalInlet` and `ExternalOutlet` are not intended for intra-streamlet communications. They should only
```
